feat: memory based connection limits

hanabi1224 opened the pull request:
## Description

<!--
Please write a summary of your changes and why you made them.
This section will appear as the commit message after merging.
Please craft it accordingly.
For a quick primer on good commit mesages, check out this blog post: https://cbea.ms/git-commit/

Please include any relevant issues in here, for example:

Related https://github.com/libp2p/rust-libp2p/issues/ABCD.
Fixes https://github.com/libp2p/rust-libp2p/issues/XYZ.
-->

Implements memory-based connection limits where the user can specify an absolute or a relative limit of the process' memory usage in relation to the available system memory.

Related: #4252.

## TODO list:

- [x] Refactor `connection_limits:Behaviour` to take a generic `ConnectionLimits` type
- [x] Rename `ConnectionLimits` to `StaticConnectionLimits`
- [x] Implement `MemoryBasedConnectionLimits`
- [ ] Disable the feature on unsupported platforms
- [x] Code docs and examples
- [x] Tests

## Notes & open questions

<!--
Any notes, remarks or open questions you have to make about the PR which don't need to go into the final commit message.
-->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

hanabi1224 committed:
feat: memory based connection limits

hanabi1224 committed:
sets system ram

thomaseizinger commented on misc/connection-limits/src/connection_limits/memory.rs:
```diff
@@ -0,0 +1,181 @@
+            log::warn!("Failed to retrive process memory stats");
+        }
+        self.last_checked = Instant::now();
+    }
```
We should do this with a proper timer that can register a waker within `NetworkBehaviour::poll`.

mxinden replied:
@thomaseizinger what would be the benefit? Say that the local node has no new connections for 5 minutes, neither inbound nor outbound. In such case there is no need to refresh the memory limit, no?

thomaseizinger replied:
> @thomaseizinger what would be the benefit? Say that the local node has no new connections for 5 minutes, neither inbound nor outbound. In such case there is no need to refresh the memory limit, no?

Yeah, if it is reasonably cheap, we should just get the latest stats upon every connection.

hanabi1224 replied:
@thomaseizinger @mxinden 
The [`memory_stats`](https://docs.rs/memory-stats/1.1.0/memory_stats/fn.memory_stats.html) function is not cheap, the [doc](https://docs.rs/memory-stats/latest/memory_stats/#caveats) says
> Getting accurate memory usage on Linux is fairly expensive and not always possible. This crate always attempts to use the statistics from [/proc/self/smaps](https://man7.org/linux/man-pages/man5/proc.5.html#:~:text=See%20user_namespaces%287%29.-,/proc/%5Bpid%5D/smaps,-%28since%20Linux%202.6.14) if avaliable. However, since support for /proc/self/smaps might not be compiled in on all kernels, this crate will also use the faster but less accurate statistics from [/proc/self/statm](https://man7.org/linux/man-pages/man5/proc.5.html#:~:text=by%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20waitpid%282%29.-,/proc/%5Bpid%5D/statm,-Provides%20information%20about) as a fallback.


> If speed is needed over accuracy, the always_use_statm feature can be enabled to always use the /proc/self/statm statistics.

I run cargo bench with below code
```rust
extern crate test;
    use test::Bencher;

    use super::*;

    #[bench]
    fn bench(bencher: &mut Bencher) {
        let  _memories = vec![1; 128 * 1024 * 1024];
        bencher.iter(|| {
            let _ = memory_stats();
        });
    }
```

`cargo +nightly bench`
```console
test bench::bench ... bench:     263,172 ns/iter (+/- 29,315)
```

`cargo +nightly bench --features always_use_statm`
```console
test bench::bench ... bench:       2,066 ns/iter (+/- 341)
```

Please let me know your suggestion. I see 3 options,
1. Periodically(say, 1s) refresh mem stats without `always_use_statm` feature.
2. Passively refresh mem stats upon every connection with a expiration(say, 1s) (The current implementation in this PR) 
3. Get the latest stats upon every connection with `always_use_statm` feature

thomaseizinger replied:
I tried to do some research on how inaccurate `/proc/self/statm` is but couldn't find much. I'd say for our purpose it would be fine? Given that it makes the code much simpler, I'd lean towards always using the fast version and simply getting them on-demand.

>  If speed is needed over accuracy, the always_use_statm feature can be enabled to always use the /proc/self/statm statistics.

This is actually bad design because if we activate this feature, every other crate in our user's dependency tree will also use this code path. I opened an issue with them: https://github.com/Arc-blroth/memory-stats/issues/6

hanabi1224 replied:
@thomaseizinger That makes sense, will switch to 3 once the issue is resolved.

hanabi1224 replied:
After thinking about it, I'm not sure if ~2000ns per function call is trivial enough and I have not tested on all platforms (Windows, OSX).  Would it be safer to still use approach 2 with a smaller refresh interval (e.g. 100ms) in case a swarm is dealing with a lot of connections?

thomaseizinger replied:
> After thinking about it, I'm not sure if ~2000ns per function call is trivial enough and I have not tested on all platforms (Windows, OSX). Would it be safer to still use approach 2 with a smaller refresh interval (e.g. 100ms) in case a swarm is dealing with a lot of connections?

Sure, let's keep the refresh approach then :)

hanabi1224 replied:
Fixed

hanabi1224 committed:
code golf

thomaseizinger commented on misc/connection-limits/src/connection_limits/memory.rs:
```diff
@@ -0,0 +1,181 @@
+}
+
+impl MemoryUsageBasedConnectionLimits {
+    pub fn new() -> Self {
```
Thinking ahead a bit, given that we might have multiple memory-based strategies in the future, I think it makes sense to design the constructor to express which strategies this is. Perhaps something like:

```rust
impl Behaviour {
	/// Configures an absolute memory threshold in bytes.
	///
	/// Once reached, no new connections will be allowed.
	pub fn with_absolute_threshold(max_bytes: usize) -> Self;

	/// Configures a memory threshold in percent, relative to the total physical memory.
	///
	/// Once reached, no more connections will be allowed.
	pub fn with_relative_threshold(percent: u8) -> Self;
}
```

This will also reduce the amount of items we need to expose in the public API. Note that I am assuming here that you never want to use those two options together in one behaviour. Users can still compose this behaviour twice if they want to do that!

hanabi1224 replied:
That makes sense. Maybe it's better to make it an enum instead of struct then

thomaseizinger replied:
For API evolvability reasons, I'd prefer it to be a struct.

thomaseizinger reviewed (commented):
Thank you for the contribution!

I'd like to discuss whether we should extend the existing `libp2p-connection-limits` here or create a new crate: `libp2p-memory-connection-limits`.

I think there a several arguments for a new crate:

- It means that we have more design freedom in regards to the API and can still release it as a patch-release because it is just additions.
- There isn't actually any shared code between those two. For example, the memory limits don't care about the connection kind at all. Tthe "put together what changes together" mantra also suggests that this should be separate.
- A dedicated `NetworkBehaviour` will be able to make use of the `poll` function to register a timer for refreshing the memory stats. Alternatively, we could simply refresh them on every connection attempt (how expensive is gathering those stats?)

mxinden committed:
fix(examples/file-sharing): set Kademlia `Mode::Server`

Explicitly set `libp2p-kad` `Kademlia::set_mode` to `Mode::Server` to reduce complexity of example, i.e. not having to explicitly set external addresses.

Pull-Request: #4197.

thomas-senechal committed:
docs: fix last OutboundQueryCompleted in kad doc

`OutboundQueryCompleted` hasn't been fully replaced by `OutboundQueryProgressed` in kad doc.

Pull-Request: #4257.

quininer committed:
fix(relay): remove unconditional `async-std` dependency

`libp2p-relay` crate specifies the `libp2p-swarm/async-std` feature, which causes an `async-std` dependency to always be introduced.

Pull-Request: #4283.

thomaseizinger committed:
deps(yamux): update yamux to `v0.12`

Pull-Request: #3013.

dependabot[bot] committed:
deps: bump Swatinem/rust-cache from 2.5.1 to 2.6.0

Pull-Request: #4288.

hanabi1224 committed:
move to new crate

move

hanabi1224 committed:
Merge remote-tracking branch 'origin/master' into memory-based-connection-limits

hanabi1224 committed:
doc

hanabi1224 committed:
lock

thomaseizinger commented on misc/memory-connection-limits/CHANGELOG.md:
```diff
@@ -0,0 +1,3 @@
+## 0.1.0
```
```suggestion
## 0.1.0 - unreleased
```

thomaseizinger commented on misc/memory-connection-limits/src/lib.rs:
```diff
@@ -0,0 +1,200 @@
+/// # use libp2p_identify as identify;
+/// # use libp2p_ping as ping;
+/// # use libp2p_swarm_derive::NetworkBehaviour;
+/// # use libp2p_memory_connection_limits as connection_limits;
```
```suggestion
/// # use libp2p_memory_connection_limits as memory_connection_limits;
```

thomaseizinger commented on misc/memory-connection-limits/src/lib.rs:
```diff
@@ -0,0 +1,200 @@
+            limits,
+            mem_tracker: ProcessMemoryUsageTracker::new(),
+        }
+    }
```
Instead of this indirection via a `MemoryUsageBasedConnectionLimits`, I'd prefer if `Behaviour` would just have two constructor functions, one for each setting. That would remove `MemoryUsageBasedConnectionLimits` from the public API.

What do you think?

hanabi1224 replied:
That's good idea, fixed

thomaseizinger commented on misc/memory-connection-limits/src/connection_limits.rs:
```diff
@@ -0,0 +1,110 @@
+#[derive(Debug, Clone, Copy)]
+pub struct MemoryUsageLimitExceeded {
+    pub process_memory_bytes: usize,
+    pub max_allowed_bytes: usize,
```
Let's make these private.

hanabi1224 replied:
Fixed

thomaseizinger commented on misc/memory-connection-limits/src/connection_limits.rs:
```diff
@@ -0,0 +1,110 @@
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "pending connections are dropped, process memory usage limit exceeded: process memory: {} bytes, max allowed: {} bytes",
```
The error that this will be embedded in already says "connection denied".

```suggestion
            "process memory usage limit exceeded: process memory: {} bytes, max allowed: {} bytes",
```

hanabi1224 replied:
Fixed

thomaseizinger commented on misc/memory-connection-limits/src/lib.rs:
```diff
@@ -0,0 +1,200 @@
+    fn refresh(&mut self) {
+        if let Some(stats) = memory_stats::memory_stats() {
+            self.physical_bytes = stats.physical_mem;
+            self.virtual_bytes = stats.virtual_mem;
```
Why not directly add them together here?

hanabi1224 replied:
When I run unit tests on windows sublinux 2 ubuntu, I see `virtual_mem` always being large, need to investigate whether it should be counted.

thomaseizinger commented on misc/memory-connection-limits/src/lib.rs:
```diff
@@ -0,0 +1,200 @@
+        _: &Multiaddr,
+        _: &Multiaddr,
+    ) -> Result<(), ConnectionDenied> {
+        self.limits.check_limit(&mut self.mem_tracker)
```
I find this API a bit odd, can we remove `ProcessMemoryUsageTracker` and only pass the total memory usage as bytes into the limits?

hanabi1224 replied:
Yes, will refactor it into approach 3 once https://github.com/Arc-blroth/memory-stats/issues/6 is done. `ProcessMemoryUsageTracker ` should no longer be needed.

> I see 3 options,
Periodically(say, 1s) refresh mem stats without always_use_statm feature.
Passively refresh mem stats upon every connection with a expiration(say, 1s) (The current implementation in this PR)
Get the latest stats upon every connection with always_use_statm feature

thomaseizinger replied:
> Yes, will refactor it into approach 3 once https://github.com/Arc-blroth/memory-stats/issues/6 is done. `ProcessMemoryUsageTracker ` should no longer be needed.
> 
> > I see 3 options,
> Periodically(say, 1s) refresh mem stats without always_use_statm feature.
> Passively refresh mem stats upon every connection with a expiration(say, 1s) (The current implementation in this PR)
> Get the latest stats upon every connection with always_use_statm feature

I'd suggest we don't block on the linked issue but implement it here by activating the feature-flag and always call the function.

It is not ideal but good enough to ship a first version. Once the new API is implemented (which I'd assume we have to contribute) then we can update the dependency and remove the usage of the feature-flag.

What do you think?

hanabi1224 replied:
That makes perfect sense, code updated

hanabi1224 replied:
Btw, I do have [some concern](https://github.com/libp2p/rust-libp2p/pull/4281#discussion_r1283411050) on its performance on other platforms like windows, I don't think the feature flag has impact on windows

thomaseizinger reviewed (commented):
Thanks! Some suggestions :)

hanabi1224 committed:
doc and tests

hanabi1224 committed:
Merge branch 'master' into memory-based-connection-limits

hanabi1224 committed:
Apply suggestions from code review

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

hanabi1224 committed:
hide MemoryUsageBasedConnectionLimits

hanabi1224 committed:
update MemoryUsageLimitExceeded fmt as suggested

hanabi1224 committed:
fix clippy

hanabi1224 committed:
fix rustdoc

thomaseizinger commented on misc/memory-connection-limits/src/lib.rs:
```diff
@@ -0,0 +1,230 @@
+    /// Updates the process memory usage threshold in the percentage of the total physical memory,
+    pub fn update_max_percentage(&mut self, percentage: f64) {
+        self.limits = MemoryUsageBasedConnectionLimits::default().with_max_percentage(percentage);
+    }
```
I think these are redundant because a user could also just assign a new instance of the entire behaviour to the existing field :)

hanabi1224 replied:
Yes, they are mostly for writing tests, see https://github.com/libp2p/rust-libp2p/pull/4281#discussion_r1285730092

thomaseizinger commented on misc/memory-connection-limits/src/lib.rs:
```diff
@@ -0,0 +1,230 @@
+            limits: MemoryUsageBasedConnectionLimits::default().with_max_percentage(percentage),
+            mem_tracker: ProcessMemoryUsageTracker::new(),
+        }
+    }
```
Slight preference towards dropping `new_` here to make the fn name shorter.

hanabi1224 replied:
Fixed

thomaseizinger commented on misc/memory-connection-limits/src/lib.rs:
```diff
@@ -0,0 +1,230 @@
+
+impl Behaviour {
+    /// Sets the process memory usage threshold in bytes,
+    /// all pending connections will be dropped when the threshold is exeeded
```
```suggestion
    /// Sets the process memory usage threshold in absolute bytes.
    ///
    /// New inbound and outbound connections will be denied when the threshold is reached.
```

Conventionally, there is a newline between heading and text.

thomaseizinger commented on misc/memory-connection-limits/src/lib.rs:
```diff
@@ -0,0 +1,230 @@
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum ConnectionKind {
```
Is this actually used?

hanabi1224 replied:
Good catch, removed

thomaseizinger commented on misc/memory-connection-limits/tests/max_bytes.rs:
```diff
@@ -0,0 +1,80 @@
+use rand::{rngs::OsRng, Rng};
+
+#[test]
+fn max_bytes() {
```
What are the chances of this test being flaky if it is run in parallel with others?

Cargo makes a new process per test file here I think so actually, we should be fine! :)

Great idea!

thomaseizinger commented on misc/memory-connection-limits/tests/util.rs:
```diff
@@ -0,0 +1,119 @@
+    fn handle_connection(&mut self) {
+        // 1MB
+        self.map.push(vec![1; 1024 * 1024]);
+    }
```
Now that is creative 😄

hanabi1224 replied:
: )

thomaseizinger commented on misc/memory-connection-limits/tests/util.rs:
```diff
@@ -0,0 +1,119 @@
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        self.handle_connection();
+        Ok(dummy::ConnectionHandler)
+    }
```
By doing this in all 4 handlers, you are allocating 2 MB per connection, I think doing it only for each established is good enough :)

hanabi1224 replied:
In these tests, connections are never established, I've also changed mem consumption for established to 0

thomaseizinger commented on misc/memory-connection-limits/tests/util.rs:
```diff
@@ -0,0 +1,119 @@
+impl ConsumeMemoryBehaviour {
+    fn handle_connection(&mut self) {
+        // 1MB
+        self.map.push(vec![1; 1024 * 1024]);
```
Let's make the 1MB configurable and pass it in from the test. I was initially confused on why the test works because there was no visible connection!

hanabi1224 replied:
I've made it a const generic and use a type alias `ConsumeMemoryBehaviour1MBPending0Established` for better readability.

```rust
let mut network = Swarm::new_ephemeral(|_| TestBehaviour {
        connection_limits: Behaviour::with_max_percentage(0.1),
        mem_consumer: ConsumeMemoryBehaviour1MBPending0Established::default(),
    });
```

thomaseizinger replied:
That works too!

thomaseizinger commented on misc/memory-connection-limits/tests/max_percentage.rs:
```diff
@@ -0,0 +1,78 @@
+
+#[test]
+fn max_percentage() {
+    let connection_limit = OsRng.gen_range(1..30);
```
No randomness please, make it a property-based test instead.

hanabi1224 replied:
This follows the [`max_outgoing`](https://github.com/libp2p/rust-libp2p/blob/f95f39f634d08dad8df72d7ecf92312f9ad2d826/misc/connection-limits/src/lib.rs#L389) test in static connection limits crate.  

And using `quickcheck` won't work here as it generates multiple tests that run in parallel (unless we supply `--test-threads 1` to `cargo test`

How about using a fixed value, say, 20?

thomaseizinger replied:
> This follows the [`max_outgoing`](https://github.com/libp2p/rust-libp2p/blob/f95f39f634d08dad8df72d7ecf92312f9ad2d826/misc/connection-limits/src/lib.rs#L389) test in static connection limits crate.

Touché! I was not aware of that :D

> How about using a fixed value, say, 20?

I'd prefer that :)

hanabi1224 replied:
Fixed

thomaseizinger commented on misc/memory-connection-limits/tests/max_bytes.rs:
```diff
@@ -0,0 +1,80 @@
+    network
+        .behaviour_mut()
+        .connection_limits
+        .update_max_bytes(max_allowed_bytes_plus_base_usage);
```
Why not initialize it directly with a higher limit? Seems more concise or am I missing something?

hanabi1224 replied:
Memory usage goes up by ~2MB after constructing the swarm, thus I need to do the calculation after Swarm being initialized

thomaseizinger replied:
Can't we initialize it with an offset of ~2MB then?

hanabi1224 replied:
I'm worried that the offset may vary in debug/release build and on different platforms. Also, a hard-coded value will likely need maintenance when `libp2p` evolves. Does that make sense?

thomaseizinger commented on misc/memory-connection-limits/src/error.rs:
```diff
@@ -0,0 +1,41 @@
+// Copyright 2023 Protocol Labs.
```
Do we need this module or can it just live in `lib.rs`?

hanabi1224 replied:
Moved to `lib.rs`

thomaseizinger reviewed (commented):
Great progress! I've added some more comments :)

hanabi1224 committed:
Update misc/memory-connection-limits/src/lib.rs

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

hanabi1224 committed:
resolve comments

hanabi1224 committed:
Make MemoryUsageLimitExceeded fields private

thomaseizinger commented on misc/memory-connection-limits/src/lib.rs:
```diff
@@ -0,0 +1,231 @@
+    /// Updates the process memory usage threshold in the percentage of the total physical memory,
+    pub fn update_max_percentage(&mut self, percentage: f64) {
+        self.max_process_memory_usage_percentage = Some(percentage);
+    }
```
Do you feel strongly about keeping these? I think they are redundant because a user can always do the following:

1. To have both limits active:

 ```rust
use libp2p_swarm_derive::NetworkBehaviour;
use libp2p_memory_connection_limits as memory_connection_limits;

#[derive(NetworkBehaviour)]
struct MyBehaviour {
   absolute_limits: memory_connection_limits::Behaviour,
   relative_limits: memory_connection_limits::Behaviour
}

fn main() {
	let _ = MyBehaviour {
		absolute_limits: memory_connection_limits::Behaviour::with_max_bytes(100000),
		relative_limits: memory_connection_limits::Behaviour::with_max_percentage(0.80),
	};
}
```

2. To update certain limits:

```rust
fn main() {
	let mut swarm = todo!();

	swarm.behaviour_mut().limits = memory_connection_limits::Behaviour::with_max_bytes(100000);
}
```

They need a `&mut`  reference to call these update functions so they can also just assign a new value to the field.

What do you think?

hanabi1224 replied:
TBH these methods are only useful for the unit tests as of now.  The unit tests are in separate files to be compiled into separate test binaries thus they have no access to Behaviour's private fields. Any suggestions?
https://github.com/libp2p/rust-libp2p/pull/4281#discussion_r1285733361

In practice, I have to call `network.dial` once before getting the process memory usage to make the tests more stable, as I found the memory usage grows  a bit more than usual after the first `network.dial` call. (Or maybe I should just sleep for 100ms to let OS refresh the stats)
```rust
// Exercise `dial` function to get more stable memory stats later
    network
        .dial(
            DialOpts::peer_id(target)
                .addresses(vec![addr.clone()])
                .build(),
        )
        .expect("Unexpected connection limit.");
```

thomaseizinger commented on misc/memory-connection-limits/src/lib.rs:
```diff
@@ -0,0 +1,231 @@
+        match (
+            self.max_process_memory_usage_bytes,
+            max_process_memory_usage_percentage,
+        ) {
```
Would it make sense to already compute `max_allowed_bytes` when the user sets the value in the constructor? i.e. only have one field in the behaviour instead of two `Option`s.

I think that would make this easier to reason about because we don't have to deal with `Option`s at all.

hanabi1224 replied:
Fixed

thomaseizinger reviewed (commented):
Great work! Two more comments :)

hanabi1224 committed:
Merge branch 'master' into memory-based-connection-limits

hanabi1224 committed:
Pre-calculate max_allowed_bytes

hanabi1224 committed:
refresh interval

thomaseizinger committed:
Don't use `update` functions in tests

thomaseizinger committed:
Remove inner `Option`

thomaseizinger committed:
Don't make refresh interval configurable

thomaseizinger committed:
Update docs

thomaseizinger committed:
Remove refresh-interval field

thomaseizinger committed:
Use early return to reduce indentation

thomaseizinger committed:
Only set timestamp if we actually refreshed the data

thomaseizinger committed:
Use early return instead of else

hanabi1224 committed:
Make tests for stable

thomaseizinger committed:
Add docs to constant

thomaseizinger committed:
Sort functions are per trait definition

thomaseizinger committed:
Merge branch 'memory-connection-limits-thomas' into memory-based-connection-limits

thomaseizinger reviewed (dismissed):
This is ready from my PoV but I've also pushed some commits to this so I'll defer to @mxinden for a final review :)

hanabi1224 committed:
Re-export from libp2p meta crate

hanabi1224 committed:
Merge branch 'master' into memory-based-connection-limits

thomaseizinger commented on libp2p/CHANGELOG.md:
```diff
@@ -6,8 +6,12 @@
   See [PR 4188].
 
+- Add `libp2p-memory-connection-limits` providing memory usage based connection limit configurations
+  See [PR 4281]
```
```suggestion
- Add `libp2p-memory-connection-limits` providing memory usage based connection limit configurations.
  See [PR 4281].
```

thomaseizinger committed:
Update libp2p/CHANGELOG.md

thomaseizinger commented on libp2p/Cargo.toml:
```diff
@@ -65,6 +66,7 @@ json = ["libp2p-request-response?/json"]
 kad = ["dep:libp2p-kad", "libp2p-metrics?/kad"]
 macros = ["libp2p-swarm/macros"]
 mdns = ["dep:libp2p-mdns"]
+memory-connection-limits = ["dep:libp2p-memory-connection-limits"]
```
Thanks for adding the feature-flag! We are adding some dependencies here so I think this is worth having. (Note that the previous two connection limits crates are not feature-flagged)

mxinden commented on misc/memory-connection-limits/tests/util.rs:
```diff
@@ -0,0 +1,128 @@
+pub(crate) type ConsumeMemoryBehaviour1MBPending0Established =
+    ConsumeMemoryBehaviour<{ 1024 * 1024 }, 0>;
+
+#[derive(Default)]
+pub(crate) struct ConsumeMemoryBehaviour<const MEM_PENDING: usize, const MEM_ESTABLISHED: usize> {
+    mem_pending: Vec<Vec<u8>>,
+    mem_established: Vec<Vec<u8>>,
+}
+
+impl<const MEM_PENDING: usize, const MEM_ESTABLISHED: usize>
+    ConsumeMemoryBehaviour<MEM_PENDING, MEM_ESTABLISHED>
+{
+    fn handle_pending(&mut self) {
+        // 1MB
+        self.mem_pending.push(vec![1; MEM_PENDING]);
+    }
+
+    fn handle_established(&mut self) {
+        // 1MB
+        self.mem_established.push(vec![1; MEM_ESTABLISHED]);
+    }
+}
```
Neat, but potentially flaky given parallel memory consumers running next to the relevant unit tests.

No change required. Let's try with this.

Thank you for investing into tests!

mxinden reviewed (dismissed):
Thank you for working on this @hanabi1224 and @thomaseizinger!

hanabi1224 committed:
Merge branch 'master' into memory-based-connection-limits
